decoder.go

```diff
@@ -20,18 +42,40 @@ func frameLineSize(frame *C.AVFrame) *C.int {
 	return (*C.int)(unsafe.Pointer(&frame.linesize[0]))
 }
 
-// h264Decoder is a wrapper around ffmpeg's H264 decoder.
-type h264Decoder struct {
-	codecCtx    *C.AVCodecContext
-	srcFrame    *C.AVFrame
-	swsCtx      *C.struct_SwsContext
-	dstFrame    *C.AVFrame
-	dstFramePtr []uint8
+// getStreamInfo opens a stream URL and retrieves its information
+func getStreamInfo(url string) (videoCodec, error) {
+	// Register all formats and codecs
+	C.avformat_network_init()
+
+	cUrl := C.CString(url)
+	defer C.free(unsafe.Pointer(cUrl))
+
+	var avFormatCtx *C.AVFormatContext = nil
+	if C.avformat_open_input(&avFormatCtx, cUrl, nil, nil) != 0 {
+		return Unknown, errors.New("avformat_open_input() failed")
+	}
+	defer C.avformat_close_input(&avFormatCtx)
+
+	if C.avformat_find_stream_info(avFormatCtx, nil) < 0 {
+		return Unknown, fmt.Errorf("avformat_find_stream_info() failed")
+	}
+
+	for i := 0; i < int(avFormatCtx.nb_streams); i++ {
+		stream := *(**C.AVStream)(unsafe.Pointer(uintptr(unsafe.Pointer(avFormatCtx.streams)) + uintptr(i)*unsafe.Sizeof(*avFormatCtx.streams)))
+		codecParams := stream.codecpar
+		if codecParams.codec_id == C.AV_CODEC_ID_H264 {
+			return H264, nil
+		} else if codecParams.codec_id == C.AV_CODEC_ID_HEVC {
+			return H265, nil
+		}
+	}
```

Comment:
1. Is it possible for a given url to support both 264 & h265?
2. If so, what should `getStreamInfo` return if the url supports both? 
3. Have we been given product requirements for that?
4. It is not clear to me that this pointer math approach is memory-safe. Could you please try writing a simple function in C in the cgo include here https://github.com/seanavery/viamrtsp/pull/1/files#diff-2a967f84f1af0d52111b8dbfbeadc44bfeb1092dacebb2bc0648a0e57284c04bR16 that takes the `avFormatCtx` which walks the datastructure & returns `-1: error, 0: unknown, 1: h264, 2: h265` while doing all the C best practice nil pointer checking, bounds checking, etc?

Reply:
Added `get_video_codec` helper fn in C .
- Removes complicated pointer logic in golang.
- Added logic to "prefer" h264 in the case that both are available.
- Did not add in an error case as there are no failure states. (We are just parsing avFormatContext->streams).